mypyc/test-data/irbuild-tuple.test

```diff
@@ -258,3 +258,222 @@ L14:
+    r9 = load_mem r8 :: int64*
+    keep_alive source
+    r10 = PyTuple_New(r9)
+    r11 = 0
+L1:
+    r12 = get_element_ptr source ob_size :: PyVarObject
+    r13 = load_mem r12 :: int64*
+    keep_alive source
+    r14 = r13 << 1
+    r15 = r11 < r14 :: signed
+    if r15 goto L2 else goto L4 :: bool
+L2:
+    r16 = CPyList_GetItemUnsafe(source, r11)
+    r17 = unbox(int, r16)
+    x = r17
+    r18 = f(x)
+    r19 = box(bool, r18)
+    r20 = CPySequenceTuple_SETITEM(r10, r11, r19)
+L3:
+    r21 = r11 + 2
+    r11 = r21
+    goto L1
+L4:
+    a = r10
+    return 1
+
+
+[case testTupleBuiltFromList]
+def f2(val: str) -> str:
+    return val + "f2"
+
+def test() -> None:
+    source = ["a", "b", "c"]
+    a = tuple(f2(x) for x in source)
+    print(a)
+[out]
+def f2(val):
+    val, r0, r1 :: str
+L0:
+    r0 = 'f2'
+    r1 = PyUnicode_Concat(val, r0)
+    return r1
+def test():
+    r0, r1, r2 :: str
+    r3 :: list
+    r4, r5, r6, r7 :: ptr
+    source :: list
+    r8 :: ptr
+    r9 :: int64
+    r10 :: tuple
+    r11 :: short_int
+    r12 :: ptr
+    r13 :: int64
+    r14 :: short_int
+    r15 :: bit
+    r16 :: object
+    r17, x, r18 :: str
+    r19 :: bit
+    r20 :: short_int
+    a :: tuple
+    r21 :: object
+    r22 :: str
+    r23, r24 :: object
+L0:
+    r0 = 'a'
+    r1 = 'b'
+    r2 = 'c'
+    r3 = PyList_New(3)
+    r4 = get_element_ptr r3 ob_item :: PyListObject
+    r5 = load_mem r4 :: ptr*
+    set_mem r5, r0 :: builtins.object*
+    r6 = r5 + 8
+    set_mem r6, r1 :: builtins.object*
+    r7 = r5 + 16
+    set_mem r7, r2 :: builtins.object*
+    keep_alive r3
+    source = r3
+    r8 = get_element_ptr source ob_size :: PyVarObject
+    r9 = load_mem r8 :: int64*
+    keep_alive source
+    r10 = PyTuple_New(r9)
+    r11 = 0
+L1:
+    r12 = get_element_ptr source ob_size :: PyVarObject
+    r13 = load_mem r12 :: int64*
+    keep_alive source
+    r14 = r13 << 1
+    r15 = r11 < r14 :: signed
+    if r15 goto L2 else goto L4 :: bool
+L2:
+    r16 = CPyList_GetItemUnsafe(source, r11)
+    r17 = cast(str, r16)
+    x = r17
+    r18 = f2(x)
+    r19 = CPySequenceTuple_SETITEM(r10, r11, r18)
+L3:
+    r20 = r11 + 2
+    r11 = r20
+    goto L1
+L4:
+    a = r10
+    r21 = builtins :: module
+    r22 = 'print'
+    r23 = CPyObject_GetAttr(r21, r22)
+    r24 = PyObject_CallFunctionObjArgs(r23, a, 0)
+    return 1
+
+
+[case testTupleBuiltFromTuple]
+def f(val: bool) -> bool:
+    return not val
+
+def test() -> None:
+    source = (True, True, False, True)
+    a = tuple(f(x) for x in source)
+[out]
+def f(val):
+    val, r0 :: bool
+L0:
+    r0 = val ^ 1
+    return r0
+def test():
+    r0, source :: tuple[bool, bool, bool, bool]
+    r1 :: list
+    r2, r3, r4 :: object
+    r5, x, r6 :: bool
+    r7 :: object
+    r8 :: int32
+    r9, r10 :: bit
+    r11, a :: tuple
+L0:
+    r0 = (1, 1, 0, 1)
+    source = r0
+    r1 = PyList_New(0)
+    r2 = box(tuple[bool, bool, bool, bool], source)
+    r3 = PyObject_GetIter(r2)
+L1:
```

Comment:
Bugs here.
I found that tuple `source` would be recognized as `tuple` instead of `builtin_tuple`, thus failed in `is_tuple_rprimitive (rtype)` in `specialize.py`

Reply:
The above type annotation should help, but it would be better if the specialization also can be made to work with fixed-length tuples.

You could add a check for `RTuple` (fixed-length tuple) and also handle that, instead of just supporting variable-length tuples and lists. You may need to coerce it to a variable-length tuple explicitly first, using `coerce`.

Reply:
Should I support `RTuple` in this PR?